0016-pre-shared-key/0016-pre-shared-key.md

```diff
@@ -0,0 +1,218 @@
+
+// This pattern of taking the HMAC of a secret and a hard-coded string will be
+// repeated for the generation of many variables. It helps secure the secrets
+// against any interactions between the algorithms used.
+
+shared_secret_generator = hmac_sha_256(receiver_secret, "ilp_psk_generation")
+shared_secret = hmac_sha_256(shared_secret_generator, token)
+
+// Now the shared_secret and the receiver_address are transmitted to the
+// sender.
+
+/** Executed by sender before each payment, to get the ILP packet and the condition.
+  * @param shared_secret (sent by the receiver)
+  * @param receiver_address (sent by the receiver)
+  *
+  * @param headers (a list of HTTP style headers, which will be encrypted)
+  * @param unsafe_headers (a list of HTTP style headers, which will be left
+  * unencrypted)
+  * @param application_data (arbitrary binary data used for higher level
+  * protocols, which will be encrypted)
+  * @param destination_amount (amount that the receiver will receive on their
+  * ledger, used to construct the ILP packet)
+  */
+
+// The private headers are put together first
+
+private_headers = create_empty_buffer()
+private_headers += headers.join('\n')
+
+// Application data is attached to the private headers, after an empty line.
+
+private_headers += '\n\n'
+private_headers += application_data
+
+// A key is created in order to encrypt the private headers and the application
+// data. A nonce is used to generate the key, so that the same private headers
+// won't be turned into the same ciphertext.
+
+nonce = random_bytes(16)
+payment_encryption_key = hmac_sha_256(shared_secret, nonce)
+encrypted_data = aes_256_ctr(payment_encryption_key, private_headers)
+
+// Now the PSK data object is constructed, containing the public headers, the
+// encrypted private headers, and the encrypted application data. A status
+// line with the PSK version is added first. The "unsafe_headers" are added
+// next, including a "key" header that specifies the HMAC algorithm used (which
+// MUST be hmac-sha-256) and the nonce used to generate the payment's
+// encryption key.
+
+psk_data = create_empty_buffer()
+psk_data += 'PSK/1.0\n'
+psk_data += 'Key: hmac-sha-256 ' + nonce
+psk_data += unsafe_headers.join('\n')
+
+// Private headers and application data are appended after an empty line.
+
+psk_data += '\n\n'
+psk_data += encrypted_data
+
+// An ILP packet is generated from the receiver's address and the amount we
+// want to send to them. The PSK data makes up the data field of the ILP
+// packet.
+
+ilp_packet = create_ilp_packet({
+  amount: destination_amount,
+  account: receiver_address,
+  data: psk_data
+})
+
+// The fulfillment is not actually used by the sender, but the condition is
+// attached to their outgoing transfer. The receiver will perform this same
+// derivation when fulfilling the payment.
+
+fulfillment_generator = hmac_sha_256(shared_secret, 'ilp_psk_condition')
```

Comment:
I think that the `PSK/1.0` should work as our version marker, because an implementation will stop parsing if the `PSK/1.0` is changed or missing. We shouldn't have to put version tags on every value we have.